### DISCUSSION TITLE
ddev with custom docker container

### DISCUSSION THREAD

**Discussion by premar** [ddev/ddev#4371] [Q&A]
Hello

To generate pdf we want to shift to a container with chromium. But the dev-team need to use the container in ddev to test too.

I created a custom _docker-compose.chromium.yaml_ inside of ddev:

```
version: '3.6'
services:
  chromium:
    container_name: ddev-${DDEV_SITENAME}-chromium
    hostname: ${DDEV_SITENAME}-chromium
    image: registry.our-company.com/docker-chromium:latest
    ports:
      - "9222"
    environment:
      - VIRTUAL_HOST=$DDEV_HOSTNAME
      - HTTP_EXPOSE=9222:9222
    labels:
      com.ddev.site-name: ${DDEV_SITENAME}
      com.ddev.approot: $DDEV_APPROOT
```

The idea is to call the container via ws://ddev-xxx-chromium:9222/devtools. The ddev stack is starting without a problem:

```
ddev start
Starting customer1... 
Container ddev-ssh-agent  Recreate 
Container ddev-ssh-agent  Recreated 
Container ddev-ssh-agent  Started 
Network ddev-customer1_default  Created 
Container ddev-customer1-dba  Started 
Container ddev-customer1-db  Started 
Container ddev-customer1-chromium  Started 
Container ddev-customer1-web  Started 
Container ddev-router  Started 
```

I can ping the chromium container from within another container:

```
ping ddev-customer1-chromium
PING ddev-customer1-chromium (172.19.0.2) 56(84) bytes of data.
64 bytes from ddev-customer1-chromium.ddev-customer1_default (172.19.0.2): icmp_seq=1 ttl=64 time=0.294 ms
(...)
```

But I cannot get access the chromium container endpoint by generated dns-name. I receive a 500 Internal Server Error. But directly with the ip its working:

```
wget http://ddev-customer1-chromium:9222/json
--2022-11-07 09:51:26--  http://ddev-customer1-chromium:9222/json
Resolving ddev-customer1-chromium (ddev-customer1-chromium)... 172.19.0.2
Connecting to ddev-customer1-chromium (ddev-customer1-chromium)|172.19.0.2|:9222... connected.
HTTP request sent, awaiting response... 500 Internal Server Error
2022-11-07 09:51:26 ERROR 500: Internal Server Error.
```

```
wget http://172.19.0.2:9222/json
--2022-11-07 09:51:42--  http://172.19.0.2:9222/json
Connecting to 172.19.0.2:9222... connected.
HTTP request sent, awaiting response... 200 OK
Length: 363 [application/json]
Saving to: ‘json.1’

json.1                                       100%[============================================================================================>]     363  --.-KB/s    in 0s      

2022-11-07 09:51:42 (32.5 MB/s) - ‘json.1’ saved [363/363]
```

I assume the ddev-router is blocking the connection. But why?

Any help would be appreciated.

**Comment by tyler36** [comment]
Hi @premar 

There are 3 Chrome impletation on DDEV-contrib that might help steer you:

- [DrupalCI ChromeDriver](https://github.com/drud/ddev-contrib/tree/master/docker-compose-services/drupalci-chromedriver) 
- [Headless Chrome Service](https://github.com/drud/ddev-contrib/tree/master/docker-compose-services/headless-chrome)
- [Puppeteer Headless Chrome support](https://github.com/drud/ddev-contrib/tree/master/recipes/puppeteer-headless-chrome-support)

The first one has the follow code which relates directly to the router but I'm not 100% sure if it will resolve your problem:

```yml
    external_links:
      - ddev-router:${DDEV_SITENAME}.${DDEV_TLD}
```

**Reply by premar** [reply]
Hi @tyler36 

Thanks for the help. At the moment only the [headless-chrome](https://github.com/drud/ddev-contrib/tree/master/docker-compose-services/headless-chrome) does what we want to achieve.

So I cloned the docker-compose from the project. But the same results:

```
wget http://ddev-customer1-chrome:9222/json
--2022-11-07 11:08:58--  http://ddev-customer1-chrome:9222/json
Resolving ddev-customer1-chrome (ddev-customer1-chrome)... 172.27.0.2
Connecting to ddev-customer1-chrome (ddev-customer1-chrome)|172.27.0.2|:9222... connected.
HTTP request sent, awaiting response... 500 Internal Server Error
2022-11-07 11:08:58 ERROR 500: Internal Server Error.
```

Unfortunately the `ddev-router:${DDEV_SITENAME}` option is not helping.

I remove the environment for the router. And added the port directly. Now there should be no router between the containers, but still 500:

```
version: '3.6'
services:
  chromium:
    container_name: ddev-${DDEV_SITENAME}-chromium
    hostname: ${DDEV_SITENAME}-chromium
    image: registry.our-company.com/docker-chromium:latest
    ports:
      - '9222:9222'
    #environment:
    #  - VIRTUAL_HOST=$DDEV_HOSTNAME
    #  - HTTP_EXPOSE=9222:9222
    labels:
      com.ddev.site-name: ${DDEV_SITENAME}
      com.ddev.approot: $DDEV_APPROOT
    #external_links:
    #  - ddev-router:${DDEV_SITENAME}
```

**Reply by rfay** [reply]
* Your 50x errors are 50x errors, so you'll have to look at things like `ddev logs -s chromium` to understand them. Please check that. 
* You can also `ddev ssh -s chromium` and explore the chromium container. Inside the container,
*  `curl -I http://chromium:9222` may be interesting (adjust the URL with a path?)
* Create a dummy project that demonstrates this and share it and we can look at it together.

**Reply by premar** [reply]
Did some debugging today. And I found the problem. I am running chromium with the following parameter:

```
chromium-browser \ 
    --disable-gpu \
    --disable-software-rasterizer \
    --headless \
    --no-sandbox \
    --allow-insecure-localhost \
    --no-first-run \
    --disable-translate \
    --disable-extensions \
    --enable-logging=stderr \
    --v=1 \
    --remote-debugging-address=0.0.0.0 \
    --remote-debugging-port=9222
```

As stated the call with IP was working, with the a hostname not. There was no log from the chromium container (even with verbose enabled).

I thought the problem was the ddev router. But your right it was the chromium all along. After some digging I found a similare problem in puppeteer: https://github.com/puppeteer/puppeteer/issues/2242

In short they disable the direct call over a hostname.

Options to work around for this problem:

1. Lookup up the IP of the Container
2. Call the endpoint directly with the IP

Apperently the *.localhost domain is also allowed, so if you rename the container hostname to `something.localhost` it also works.

I am not happy with neither of those options. But if ever someone is searching for the same problem.

**Reply by rfay** [reply]
You may want to take a look at the new add-on https://github.com/weitzman/ddev-selenium-standalone-chrome

**Reply by premar** [reply]
@rfay thanks for the hint, but that would be a little overhead for simply generating pdf's? Our goal ist to use the https://github.com/chrome-php/chrome wrapper to open the chrome remote debug endpoint. There generate the pdf of a given website, and save the result.

And I am not sure if you can call the remote debug endpoint inside of the selenium container. I am going to tweak my container and then publish it later if it works as expected.

But thats all not a problem of ddev. For me the problem is resolved, because it was not ddev in the first place.

**Reply by rfay** [reply]
You're just trying to generate PDFs? Why don't you try a maintained solution like https://github.com/drud/ddev-pdfreactor ? Or one of many other solutions? Anyway, glad the problem is resolved.
